File: src/numeric/euclidgcd/euclidgcd.pas
{
    Calculates the GCD of two values and prints it all to the screen
}

program euclidgcd;
    uses SysUtils;

    (* Calculate the GCD of the two numbers with Euclid's algorithm *)
    function gcd(m, n : integer) : integer;
        var r : integer;
    begin
        while n <> 0 do begin
            r := m mod n;
            m := n;
            n := r
        end;
        gcd := m
    end;

    var m, n, value : integer;
begin
    // try to get the first 2 command line arguments or use 15, 10
    if ParamCount >= 2 then begin
        m := StrToInt(ParamStr(1));
        n := StrToInt(ParamStr(2))
    end
    else begin
        m := 15;
        n := 10;
    end;

    value := gcd(m, n);
    WriteLn(m, ' ', n);
    WriteLn('gcd: ', value)
end.

File: src/numeric/euclidgcd/euclidgcd.ml
(* Calculate the GCD between two values and print it all to the screen *)

(* Get the GCD via recrusive form of Euclid's algorithm *)
let rec euclidgcd m n = 
  if n = 0 then m else euclidgcd n (m mod n);;

(* Parse a command line argument as an integer; or return default *)
let argAsInt index default =
  if (Array.length Sys.argv) >= 3 then
    (int_of_string Sys.argv.(index)) else default;;

(* Get the command line arguments, or 15, 10 *)
let m = argAsInt 1 15 in
let n = argAsInt 2 10 in
let result = euclidgcd m n in
Printf.printf "%d %d\ngcd: %d\n" m n result;;
